sgl-kernel/csrc/cpu/norm.cpp

```diff
@@ -304,24 +304,33 @@ void fused_rmsnorm_gated_kernel_impl(
 
 template <typename scalar_t>
 void fused_add_layernorm_kernel_impl(
-    scalar_t* __restrict__ input,
+    scalar_t* __restrict__ output,
+    const scalar_t* __restrict__ input,
     scalar_t* __restrict__ residual,
     const scalar_t* __restrict__ weight,
     float* __restrict__ buffer,
     int64_t batch_size,
+    int64_t seq_len,
     int64_t hidden_size,
     int64_t input_strideN,
     float eps = 1e-5) {
   using bVec = at::vec::Vectorized<scalar_t>;
   using fVec = at::vec::Vectorized<float>;
+  TORCH_CHECK(
+      !((residual != nullptr) && (buffer == nullptr)),
+      "fused_add_layernorm_kernel_impl with a residual input requires a buffer.");
 
   constexpr int kVecSize = bVec::size();
-  at::parallel_for(0, batch_size, 0, [&](int64_t begin, int64_t end) {
-    int tid = at::get_thread_num();
-    float* __restrict__ buffer_ptr = buffer + tid * hidden_size;
+  int64_t parallel_size{(seq_len == 0) ? batch_size : batch_size * seq_len};
+  at::parallel_for(0, parallel_size, 0, [&](int64_t begin, int64_t end) {
+    float* __restrict__ buffer_ptr{(float*)nullptr};
+    if (buffer != nullptr) {
+      buffer_ptr = buffer + at::get_thread_num() * hidden_size;
+    }
 
     for (int64_t i = begin; i < end; ++i) {
-      scalar_t* __restrict__ input_ptr = input + i * input_strideN;
+      scalar_t* __restrict__ out_ptr = output + i * hidden_size;
+      const scalar_t* __restrict__ input_ptr = input + i * input_strideN;
       scalar_t* __restrict__ residual_ptr{(scalar_t*)nullptr};
       if (residual != nullptr) {
         residual_ptr = residual + i * hidden_size;
```

Comment:
usually i would prefer to use a template parameter 

<template typename has_residual>

something like that,

and use `AT_DISPATCH_BOOL` to skip the bool checks